% \VignetteIndexEntry{maSigPro Vignette}
% \VignetteDepends{maSigPro}
% \VignetteKeywords{Expression Analysis, Time course}
% \VignettePackage{maSigPro}

\documentclass[a4paper,11pt]{article}
\usepackage{Sweave}
\usepackage{url}
\usepackage{amsmath,pstricks,fullpage}
\usepackage[authoryear,round]{natbib}
\usepackage{hyperref}
\usepackage{a4wide}
\usepackage[T1]{fontenc}
\parindent 0in

\begin{document}

\title{maSigPro: Analysis of gene expression Significant Profiles}
\author{ Mar\'ia J. Nueda$^{1}$ and Ana Conesa$^{2,3}$}
\date{11 May 2017}
\maketitle

\begin{center}
$^{1}$Mathematics Department, University of Alicante, Spain\\
{\tt mj.nueda@ua.es}\\
$^{2}$ Genomics of Gene Expression Laboratory, Centro de Investigaciones Pr\'incipe Felipe, Valencia, Spain \\
{\tt aconesa@cipf.es}\\
$^{3}$ Microbiology and Cell Science Department, Institute for Food and Agricultural Research, University of Florida, USA
\end{center}

{\bf maSigPro} is a R package for the analysis of single and multiseries time course microarray and RNA-Seq experiments.
\\
{\bf maSigPro} follows a two steps regression strategy to find genes with significant temporal expression changes and significant differences between experimetal groups.
The method firstly defines a general regression model for the data where the experimental groups are identified by dummy variables.
The procedure adjusts this global model by the least squared technique to identify differentially expressed genes and selects significant genes aplying false discovery rate control procedure.
Secondly, stepwise regression is applied as a variable selection strategy to study differences between experimental groups and to find statistically significant different profiles.
The coefficients obtained in the second regression model will be useful to cluster together significant genes with similar expression patterns and to visualize the results.
\\
{\bf maSigPro} also includes several tools for the analysis of alternative isoform expression in time
course transcriptomics experiments. 
\\
To obtain the User's Guide you need to install the maSigPro package. Type at the R prompt:


\begin{Schunk}
\begin{Sinput}
> library(maSigPro)
> maSigProUsersGuide()
\end{Sinput}
\end{Schunk}

\end{document}
