---
title: "scMaSigPro: Quick Start Guide"
subtitle: "Basic Steps"
author: "Priyansh Srivastava"
package: "scMaSigPro"
abstract: |
  `scMaSigPro` is an R package designed for the serial analysis of single-cell 
  RNA-seq (scRNA-Seq) data along inferred pseudotime. It builds upon the [`maSigPro`](https://www.bioconductor.org/packages/release/bioc/html/maSigPro.html)
  Bioconductor package to identify genes with significant expression changes across different 
  branching paths in a pseudotime-ordered scRNA-Seq dataset. This vignette illustrates
  the basic workflow of `scMaSigPro`, providing a step-by-step guide for users using a
  small simulated dataset.
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{scMaSigPro Basic Workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, "setup", include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(scMaSigPro)
```

## Introduction
`scMaSigPro` is a polynomial regression-based approach inspired by the [maSigPro](https://www.bioconductor.org/packages/release/bioc/html/maSigPro.html)
Bioconductor package tailored for scRNA-Seq data. It first discretizes single cell
expression along inferred pseudotime while preserving order. Afterwards it applies
the [maSigPro model](https://academic.oup.com/bioinformatics/article/22/9/1096/200371)
to pinpoint genes exhibiting significant expression profile differences among
branching paths and pseudotime.

## Installation
Currently, `scMaSigPro` is available on GitHub and can be installed as follows:

```{r, echo=TRUE, eval=FALSE}
# Install devtools if not already installed
if (!requireNamespace("devtools", quietly = TRUE)) {
  install.packages("devtools")
}

# Install scMaSigPro
devtools::install_github("BioBam/scMaSigPro",
  ref = "main",
  build_vignettes = FALSE,
  build_manual = TRUE,
  upgrade = "never",
  force = TRUE,
  quiet = TRUE,
)
```

## Basic Workflow
Here, we demonstrate the basic workflow of `scMaSigPro` using a simulated dataset
included in the package, simulated via [splatter](https://bioconductor.org/packages/release/bioc/html/splatter.html).

### Load the `scMaSigPro` package and the dataset
```{r, "Step 1: Load the `scMaSigPro` package and the dataset", eval=TRUE, echo=TRUE}
# Set Seed for Reproducibility
set.seed(123)

# Load Package
library(scMaSigPro)

# Load example data
data("splat.sim", package = "scMaSigPro")
```

####  Plot the data (Optional)
```{r, "Plot the data (Optional)", eval=TRUE, echo=TRUE}
# Load
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(SingleCellExperiment))

# Extract PCA
df <- as.data.frame(reducedDim(splat.sim)[, c(1:2)])
df$Step <- splat.sim@colData$Step
df$Group <- splat.sim@colData$Group

# Plot the data
ggplot(
  df,
  aes(x = PC1, y = PC2, color = Step, shape = Group)
) +
  geom_point() +
  theme_minimal()
```

If you are interested in learning how this data is simulated and what information
it contains, you can try `?splat.sim` in the R console to read about the details.

The UMAP visualization of our dataset reveals a bifurcating trajectory,
illustrating how one cell type diverges into two distinct cell types. This 
bifurcation is driven by the selective expression of genes along the pseudotime.
In the UMAP plot, the two branching paths of differentiation are represented 
by points with distinct shapes, each corresponding to a different cell type.

Furthermore, the points in the UMAP are color-coded based on a simulated
pseudotime-like variable, termed _Step_, from the [Splatter package](https://bioconductor.org/packages/release/bioc/html/splatter.html). This
color gradient effectively captures the progression of cells along the pseudotime
axis, providing insights into the temporal dynamics of cellular differentiation.

### Convert to `scMaSigPro` Object
`scMaSigPro` offers various options to directly convert widely used single-cell
related S4 objects directly to the scmpClass class (now referred to as `scmpObject`),
such as [`SingleCellExperiment`](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html) or [`Monocle3`](https://cole-trapnell-lab.github.io/monocle3/)'s [`cell_data_set`](https://rdrr.io/github/cole-trapnell-lab/monocle3/man/new_cell_data_set.html) class. Here, using the
simulated object (of the [`SingleCellExperiment`](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html) class) that we have just loaded into the R environment, we will 
demonstrate the direct conversion to `scmpObject` object.

In our case, we already have pseudotime and path information stored in the
`colData` slot of the `splat.sim`. The _Step_ column holds the information for the
simulated steps of each path, which we can treat as a pseudotime variable, and
the _Group_ column has information regarding branching path, i.e., to which
branching path each of the cell belongs. We can view this information by executing 
`View(as.data.frame(splat.sim@colData)).`

```{r, "View(as.data.frame(splat.sim@colData))", eval=TRUE, echo = FALSE}
print(head(as.data.frame(splat.sim@colData), 5))
```

We will use `as_scmp()` from `scMaSigPro` and direct it to use the columns present in
the dataset as the columns for `Pseudotime` and `Path`. To do this, we have to enable
the parameter `labels_exist` and then pass the existing column names as a named list.

```{r, "Convert to `scMaSigPro` Object", eval=TRUE, echo=TRUE}
# Helper Function to convert annotated SCE object to scmpObject
scmp_ob <- as_scmp(
  object = splat.sim, from = "sce",
  align_pseudotime = FALSE,
  verbose = TRUE,
  additional_params = list(
    labels_exist = TRUE,
    exist_ptime_col = "Step",
    exist_path_col = "Group"
  )
)
```

Once the object is created, we can type the name of the object in the R console
to view various attributes, such as the dimensions (number of cells and number
of genes), the available branching paths, and the range of pseudotime.

```{r, "Console Echo-1", eval=TRUE, echo = FALSE}
scmp_ob
```

### Pseudo-bulking along the continuum with `sc.squeeze()`

`scMaSigPro` provides a comprehensive function for pseudo-bulking along the
pseudotime continuum, named `sc.squeeze()`. This function discretizes the original 
pseudotime values into bins using histogram binning methods. The `sc.squeeze()`
function includes various parameters that can be tailored depending on the 
characteristics of the data. Here, we will show the basic usage with the default
parameters:

```{r, "Pseudo-bulking along the continuum with `sc.squeeze()`",eval=TRUE, echo=TRUE}
scmp_ob <- sc.squeeze(scmp_ob)
```

The console output of `scMaSigPro` is dynamic and displays more attributes as the
analysis progresses. To view the results of the binning procedure, we can simply
type the object's name into the console:

```{r,  "Console Echo-2", eval=TRUE, echo = FALSE}
scmp_ob
```
Here, we observe that the original pseudotime, which ranged from 1 to 100 with
two cells at each step and different paths, is now rescaled to a range of 1 to 8,
with each path having an equal number of bins (8+8 in total). We also see the average
bin sizes per path, indicating the average number of cells that make up a single
bin for each path.

### Visualize bins

We can also visually inspect the binning process using a tile plot:

```{r, "Visualize bins",eval=TRUE, echo=TRUE}
plotBinTile(scmp_ob)
```

The tile plot provides a clear view of the number of cells in each bin and how the
sizes of bins compare with those of other paths in the same scmp_binned_pseudotime.

### Setting up the Polynomial Model

`scMaSigPro`, being the successor of [`maSigPro`](https://www.bioconductor.org/packages/release/bioc/html/maSigPro.html), utilizes the same polynomial regression model. Let's consider a case with two branching paths along the same
pseudotime scale, modeled with a quadratic polynomial. The model can be represented
as follows:

```{r, "formula", echo=FALSE}
library(knitr)
knitr::asis_output("
\\begin{align*}
Y_{i} &\\sim \\text{NegativeBinomial}(\\mu_{i}, \\theta = 10) \\\\
\\log(\\mu_{i}) &= \\beta_{0} + \\beta_{1} \\cdot (\\text{Path}_{B}\\text{vsPath}_{A})_{i} + \\beta_{2} \\cdot T_{\\text{Pseudotime}}^{\\text{Binned}_{i}} \\\\
&\\quad + \\beta_{3} \\cdot (T_{\\text{Pseudotime}}^{\\text{Binned}_{i}} \\cdot \\text{Path}_{B_{i}})  + \\beta_{4} \\cdot (T_{\\text{Pseudotime}}^{\\text{Binned}_{i}})^2 \\\\
&\\quad + \\beta_{5} \\cdot ((T_{\\text{Pseudotime}}^{\\text{Binned}_{i}})^2 \\cdot \\text{Path}_{B_{i}}) + \\text{Offset}_{i} + \\omega_{i} \\cdot \\epsilon_{i}\\\\
\\end{align*}
")
```

To construct this model, we use `sc.set.poly()` to include quadratic terms:

```{r, "Polynomial Degree 2",eval=TRUE, echo=TRUE}
# Polynomial Degree 2
scmp_ob <- sc.set.poly(scmp_ob, poly_degree = 2)
```

#### Visualize the model

Once the model is stored, we can visualize the corresponding polynomial using the `showPoly()` function:

```{r, "showPoly",eval=TRUE, echo=TRUE}
showPoly(scmp_ob)
```

Similarly, we can fit a cubic polynomial by setting the polynomial degree to 3:

```{r, "Polynomial Degree 3",eval=TRUE, echo=TRUE}
# Polynomial Degree 3
scmp_ob <- sc.set.poly(scmp_ob, poly_degree = 3)
showPoly(scmp_ob)
```

However, for simplicity, we will explore a polynomial of degree 1. Note that 
increasing the polynomial degree enhances `scMaSigPro`'s performance in capturing
exponential and nonlinear gene expression patterns:

```{r, "# Polynomial Degree 1", eval=TRUE, echo=TRUE}
# Polynomial Degree 1
scmp_ob <- sc.set.poly(scmp_ob, poly_degree = 1)
showPoly(scmp_ob)
```

In the above model we have:

* _beta0_: Accounts for differences in expression from the start to the end.

* _beta1*Path2vsPath1_: Captures differences between the branching paths,
assuming path-1 does not change along pseudotime.
    
* _beta2*scmp_binned_pseudotime_: Reflects differences across pseudotime.

* _beta3*scmp_binned_pseudotimexPath2_: Represents interaction between pseudotime
and path differences.

### Detecting Genes with Non-Flat Profiles

To identify genes that demonstrate significant changes along the pseudotime contniuum,
we use the `sc.p.vector()`. This function, adapted from the original [`maSigPro`](https://www.bioconductor.org/packages/release/bioc/html/maSigPro.html),
includes additional parameters such as the use of `offsets` and `weights`.

`scMaSigPro` expects raw counts as input because it models data using a negative
binomial distribution, a count distribution, so the counts should not be normalized
to continuous values. To account for library size differences, it uses `offsets`,
which are the log of size factors, similar to DESeq2.

For those who prefer to supply continuous data normalized using methods like log(1+x),
the distribution can be changed during model fitting.

We can execute `sc.p.vector()` as follows:

```{r, "Detecting Genes with Non-Flat Profiles", eval=TRUE, echo=TRUE}
# Detect non-flat profiles
scmp_ob <- sc.p.vector(scmp_ob,
  offset = TRUE, p_value = 0.05, verbose = FALSE,
  log_offset = TRUE
)
scmp_ob
```

The console output reveals that `scMaSigPro` detected 51 genes with non-flat 
profiles.

### Model Refinement

Having identified genes with significant profiles, we can refine their polynomial
models using `sc.t.fit()`. This function evaluates each term of the polynomial 
model. In our case, it will assess which among "beta0 + beta1\*Path2vsPath1 + 
beta2\*scmp_binned_pseudotime + beta3*scmp_binned_pseudotimexPath2" 
significantly contributes to the differences. To execute `sc.t.fit()`, 
we proceed as follows:

```{r, "Model Refinement",eval=TRUE, echo=TRUE}
# Model refinement
scmp_ob <- sc.t.fit(scmp_ob, verbose = FALSE)
scmp_ob
```

---

## Selection of Genes

With our refined models in hand, we now focus on identifying genes showing 
significant differences with pseudotime, among paths, or both. For this purpose,
we use the `sc.filter()` function. Our aim is to select models with a 
relatively high $R^2$, indicating simple linear relationships. The `vars` 
parameter in `sc.filter()` allows us to extract different sets of 
significant genes. Setting `vars = 'all'` retrieves all non-flat profiles 
identified in `sc.p.vector()` with $R^2>=$ the specified threshold. The option
`vars = 'groups` fetches genes per path, resulting in two gene lists that 
demonstrate associative significance among paths, helping us identify genes 
associated with one path or the other along the pseudotime continuum. The 
`vars = 'each'` option finds significance for each term in the polynomial. 
In our case, we are interested in genes differentially expressed between paths
and over pseudotime continuum, so we will choose `vars = 'groups`.
```{r, "vars = groups",eval=TRUE, echo=TRUE}
scmp_ob <- sc.filter(
  scmpObj = scmp_ob,
  rsq = 0.7,
  vars = "groups",
  intercept = "dummy",
  includeInflu = TRUE
)
```

By setting the vars parameter to "groups", the function will add genes with
$R^2$ >= 0.7 to the object. To explore the number of genes per group, we will 
make an upset plot:
```{r, "uspet",eval=TRUE, echo=TRUE}
plotIntersect(scmp_ob)
```

Here, we observe that 23 genes belong to both Path2vsPath1 and Path1, indicating
that these genes not only change along pseudotime but also exhibit significantly
different expression between the two paths. Additionally, there are 10 genes
uniquely associated with Path2vsPath1. This implies that Path2 has 10 genes that
are significantly differentially expressed over time, using Path1 genes as a
reference. Let's explore a few of these genes:
```{r, "trend",eval=TRUE, echo=TRUE}
FigureA <- plotTrend(scmp_ob, "Gene9", logs = TRUE, logType = "log")
FigureB <- plotTrend(scmp_ob, "Gene95", logs = TRUE, logType = "log")
FigureC <- plotTrend(scmp_ob, "Gene10", logs = TRUE, logType = "log")
FigureD <- plotTrend(scmp_ob, "Gene92", logs = TRUE, logType = "log")
(FigureA + FigureB) / (FigureC + FigureD)
```

These plots illustrate the gene expression trends for selected genes, providing
insights into their behavior in different paths. Fortunately, since we are using
simulated data, we can validate these trends by extracting gene-level metadata
from our simulated object:
```{r, "ground truth",eval=TRUE, echo=TRUE}
groundTruth <- as.data.frame(splat.sim@rowRanges@elementMetadata)
print(groundTruth[groundTruth$Gene %in% c("Gene9", "Gene95", "Gene10", "Gene92"), c(1, 2, 6, 8)])
```

The ground truth data reveals, for example, that Gene9 has a base gene mean of
0.5, a fold change of 1.5 in Path1, and remains at the same expression level for
Path2. This trend is accurately captured in our analysis, as shown in Figure-A.
Similarly, Gene92 does not show a significant difference between the paths but 
demonstrates a downtrend along pseudotime in both paths, a finding we 
successfully recapitulate in Figure-D.

### Cluster Trends
To discern general trends among the genes, we can cluster and visualize them 
together using the `sc.cluster.trend()` and `plotTrendCluster()`. By default, 
this function employs hierarchical clustering (hclust) and we will divide the 
genes into 4 clusters.
```{r, "Cluster Trends"}
# Cluster Trend
scmp_ob <- sc.cluster.trend(
  scmp_ob,
  geneSet = "union",
  k = 4
)
```

```{r, "plot"}
# Plot
plotTrendCluster(
  scmpObj = scmp_ob,
  plot = "coeff",
  logs = TRUE,
  verbose = FALSE
)
```

This visualization helps in understanding the collective behavior of genes
within each cluster, highlighting patterns and trends that might be relevant for
further biological interpretation.

This concludes the basic usage quick start guide of `scMaSigPro`. Please refer 
to other vignettes for more in-depth analysis.

---

### Session Info
```{r, "Session Info"}
sessionInfo(package = "scMaSigPro")
```
