---
title: "scMaSigPro: Bridging maSigPro Analysis"
subtitle: "Guide to applying maSigPro workflow with scMaSigPro"
author: "Priyansh Srivastava"
package: "scMaSigPro"
abstract: |
  `scMaSigPro` is an R package designed for the serial analysis of single-cell 
  RNA-seq (scRNA-Seq) data along inferred pseudotime. It builds upon the [`maSigPro`](https://www.bioconductor.org/packages/release/bioc/html/maSigPro.html)
  Bioconductor package to identify genes with significant expression changes across different 
  branching paths in a pseudotime-ordered scRNA-Seq dataset. This vignette illustrates
  the workflow of the original [`maSigPro`](https://www.bioconductor.org/packages/release/bioc/html/maSigPro.html) Bioconductor package of using `scMaSigPro`. 
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{MaSigPro workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, "setup", include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(scMaSigPro)
```

## Introduction
`scMaSigPro` is a polynomial regression-based approach inspired by the [maSigPro](https://www.bioconductor.org/packages/release/bioc/html/maSigPro.html)
Bioconductor package tailored for scRNA-Seq data. It first discretizes single cell
expression along inferred pseudotime while preserving order. Afterwards it applies
the [maSigPro model](https://academic.oup.com/bioinformatics/article/22/9/1096/200371)
to pinpoint genes exhibiting significant expression profile differences among
branching paths and pseudotime.

## Overview of `maSigPro`
maSigPro (microarray significant profiles) is a bioconductor package. Originally
designed for microarray data analysis, maSigPro focuses on identifying genes
with significant expression profiles that vary among different experimental
groups, particularly in time-course experiments.

## Adaptation for Bulk-RNA Seq Data
In 2016, the maSigPro package was adapted for bulk-RNA seq data, a crucial
development given the distinct differences between microarray and RNA-Seq
technologies. Unlike the distributions typically seen in microarray data,
bulk-RNA Seq data generally follows a Negative Binomial Distribution, which is
more apt for count data like RNA-Seq, characterized by variance often exceeding
the mean. 

This adaptation brought in explicit theta parameterization, critical
for RNA-Seq data analysis as the theta parameter represents the dispersion in
the Negative Binomial Distribution. Such parameterization enhances the accuracy
in modeling gene expression variability, a key factor in identifying significant
changes in gene expression, thereby marking a significant step forward in the
application of `maSigPro` to the more complex RNA-Seq data.

## Setup and Installation
### Installing `scMaSigPro`
`scMaSigPro` is currently available on GitHub. Follow these steps to install it:

```{r, "scMaSigPro-install", echo=TRUE, eval=FALSE}
# Use public PAT
publicPat <- "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"

# Install devtools if not already installed
if (!requireNamespace("devtools", quietly = TRUE)) {
  install.packages("devtools")
}

# Install scMaSigPro
devtools::install_github("spriyansh/scMaSigPro",
  ref = "dev",
  auth_token = publicPat,
  build_vignettes = FALSE,
  build_manual = TRUE,
  upgrade = "never",
  force = TRUE,
  quiet = TRUE,
)
```

### Installing `maSigPro`
You can easily install `maSigPro` from Bioconductor using the following steps:

```{r, "maSigPro-install", echo=TRUE, eval=FALSE}
# Install Bioconductor if not already installed
if (!require("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

# install maSigPro
BiocManager::install("maSigPro")
```

### Setup
First, we need to load both `scMaSigPro` and `maSigPro`. Additionally, we will load
`testthat`, a user-friendly testing framework for R that integrates well with
existing workflows. We will use testthat to verify if `scMaSigPro` produces 
the same results as `maSigPro` when used in bulk mode.
```{r, "set up", echo=TRUE, eval=TRUE}
# load scMaSigPro
library(scMaSigPro)

# load maSigPro
library(maSigPro)

# We will also keep checking the output
library(testthat)
```

### Setting up the data
In this tutorial, we will use datasets from `maSigPro` and aim to replicate the 
same analysis pipeline using `scMaSigPro`. For more detailed information about
the data analysis, refer to the original [`maSigPro` vignette-userguide](https://www.bioconductor.org/packages/release/bioc/vignettes/maSigPro/inst/doc/maSigProUsersGuide.pdf)

```{r, "get-data", echo=TRUE, eval=TRUE}
# load counts
data(data.abiotic)

# load Design file
data(edesign.abiotic)

# save with different names for scMaSigPro
count <- data.abiotic
metadata <- as.data.frame(edesign.abiotic)
```

Let's briefly examine each dataset:
```{r, "view-data", echo=TRUE, eval=TRUE}
# View a subset of count data
print(data.abiotic[c(1:5), c(1:5)])

# View a subset of the design file
print(edesign.abiotic[c(1:5), c(1:5)])
```

In edesign.abiotic, there are four groups, including one control group. The
experiment features experimental capture times, which are distinct from Pseudotime,
and includes temporal replicates.

To simplify the analysis, we'll convert the binarized groups into a single column.
`scMaSigPro` automatically binarizes groups internally, reducing the likelihood of errors.
```{r, "non-binarize", echo=TRUE, eval=TRUE}
# Add group column
metadata$Group <- apply(metadata[, c(3:6)], 1, FUN = function(x) {
  return(names(x[x == 1]))
})

# Remove binary columns
metadata <- metadata[, !(colnames(metadata) %in%
  c("Control", "Cold", "Heat", "Salt")),
drop = FALSE
]

# View
print(metadata[c(1:5), ])
```

## Creating objects for `scMaSigPro`
`scMaSigPro` utilizes S4 classes, and therefore, the initial step involves creating
an object of the `scMaSigPro` class, also known as scmpObject. To do this, we use
the `create.scmp()` function. We will use the same object throughout the tutorial
to show the concept of S4 classes. On the other hand for `maSigPro` we will be making
different S3 objects as shown in the [`original tutorial.](https://www.bioconductor.org/packages/release/bioc/vignettes/maSigPro/inst/doc/maSigProUsersGuide.pdf) 

As we are using `scMaSigPro` in bulk mode, we will set use_as_bin = TRUE. This 
directs the data to the dense slot of the scmpObject, allowing us to proceed with
the analysis without using `sc.squeeze()`, which is specifically designed for scRNA-Seq
pseudo-bulking.
```{r, "create-scmpObject", echo=TRUE, eval=TRUE}
scmp.ob <- create.scmp(
  counts = count,
  cell_data = metadata,
  pseudotime_colname = "Time",
  path_colname = "Group",
  use_as_bin = TRUE
)

# Print Console output
scmp.ob
```

In the console output, we can observe that the data has been transferred to the
dense slot. We can visualize this using `plotBinTile()`.

```{r, "plot-bin-tile"}
plotBinTile(scmp.ob)
```
 
Next, we need to verify if `scMaSigPro` has created the appropriate count matrix.
For this, we will use `testthat::expect_identical()`. This function highlights any 
mismatches; if everything is identical, it will not throw an error. 
The `eDense()` function retrieves the dense expression file from the `scMaSigPro` object.
 
```{r, "eDense(scmp.ob)"}
expect_identical(eDense(scmp.ob), expected = data.abiotic)
```

No errors were thrown, indicating that our setup is correct and we can proceed
with the analysis.

## Creatin of Regression Matrix 
The function `maSigPro::make.design.matrix()` is used to create a regression matrix
for fitting the full regression model. In `scMaSigPro`, this task is accomplished
using `sc.set.poly()`, where the regression matrix is referred to as "predictor".
We will also verify that `scMaSigPro` has correctly performed binarization, as
specified in `edesign.abiotic`.

### `maSigPro::make.design.matrix()` and `scMaSigPro::sc.set.poly()`
```{r, "make.design.matrix"}
# Using maSigPro
design <- make.design.matrix(edesign.abiotic, degree = 2)

# Using scMaSigPro
scmp.ob <- sc.set.poly(scmp.ob, poly_degree = 2)

# Comparing binarization results
expect_equal(bAlloc(scmp.ob), expected = edesign.abiotic)

# Comparing regression matrices
expect_identical(scmp.ob@design@predictor, expected = design$dis)
```

No errors were thrown, so we can proceed with the analysis.

## Detecting Non-Flat Profiles with `maSigPro::p.vector()`
To compute a regression fit for each gene, `maSigPro` uses `p.vector()`, while `scMaSigPro`
uses `sc.p.vector()`. The main difference is the offset parameter in `scMaSigPro`,
which accounts for variations in bin sizes. Since we are not dealing with
scRNA-Seq data, we will disable this parameter.
```{r, "compare p.vetor"}
# Using maSigPro
gc <- capture_output(fit <- p.vector(data.abiotic, design,
  Q = 0.05,
  MT.adjust = "BH", min.obs = 20
))

# Using ScMaSigPro
scmp.ob <- sc.p.vector(scmp.ob,
  min.na = 20,
  verbose = FALSE,
  offset = FALSE,
  max_it = 25,
  epsilon = 0.00001,
  family = gaussian()
)
```

Now, we'll compare the S3 object `fit from `maSigPro::p.vector()` with the 
results from `sc.p.vector()`:

```{r, "testthat pvector"}
# Compare p-values
expect_identical(
  matrix(scmp.ob@profile@p.vector,
    dimnames = list(names(scmp.ob@profile@p.vector), "p.value")
  ),
  expected = as.matrix(fit$p.vector[, 1, drop = FALSE])
)

# Compare adjusted p-values
pad <- scmp.ob@profile@p.adjusted
names(pad) <- NULL
expect_identical(pad, expected = fit$p.adjusted)
```

## Detecting Significant Polynomial Terms with `maSigPro::T.fit()`
Next, we apply a variable selection procedure to identify significant polynomial
terms for each gene, using `T.fit()` and `sc.t.fit()`.
```{r, "run sc.t.fit"}
# Using maSigPro
gc <- capture_output(tstep <- T.fit(fit, step.method = "backward", alfa = 0.05))

# Using scMaSigPro
scmp.ob <- sc.t.fit(scmp.ob,
  offset = FALSE,
  verbose = FALSE,
  epsilon = 0.00001,
  family = gaussian()
)
```

We will compare the s3 object `tstep` from `maSigPro::T.fit()` with the results from `sc.t.fit()`:

```{r, "comapte sc.t.fit"}
# Solutions
expect_identical(showSol(scmp.ob), expected = tstep$sol)

# Coefficients
expect_identical(showCoeff(scmp.ob), expected = tstep$coefficients)

# Group Coefficients
expect_identical(showGroupCoeff(scmp.ob), expected = tstep$group.coeffs)
```

## R-Square Filter with `maSigPro::get.siggenes()`
The next step is to generate lists of significant genes based on desired biological question.
This is achieved using the `get.siggenes()` function in `maSigPro` and `sc.filter()` in `scMaSigPro`.

```{r, "RUn r2 filter"}
# maSigPro
sigs <- get.siggenes(tstep, rsq = 0.6, vars = "groups")

# scMASigPro
scmp.ob <- sc.filter(scmp.ob, rsq = 0.6, vars = "groups")
```

We'll compare the s3 object (sigs) from `maSigPro::get.siggenes()' with the results from sc.filter()':

```{r}
# Compare Cold vs Control
expect_identical(scmp.ob@sig.genes@sig.genes$ColdvsControl,
  expected = sigs$summary$ColdvsControl
)

# Compare Heat vs Control
expect_identical(scmp.ob@sig.genes@sig.genes$HeatvsControl,
  expected = sigs$summary$HeatvsControl[sigs$summary$HeatvsControl != " "]
)

# Compare Salt vs Control
expect_identical(scmp.ob@sig.genes@sig.genes$SaltvsControl,
  expected = sigs$summary$SaltvsControl[sigs$summary$SaltvsControl != " "]
)

# Compare Control
expect_identical(scmp.ob@sig.genes@sig.genes$Control,
  expected = sigs$summary$Control[sigs$summary$Control != " "]
)
```

All comparisons indicate equality, confirming that the results are reproducible
up to this point.

## Visualizing the trends and Intersection

Having confirmed that the numerical values are identical for both methods, 
our next step is to assess whether the actual trends are also consistent.

### Visualizing Intersection `maSigPro::suma2Venn()`
```{r, "upset and venn"}
# Venn Diagram of maSigPro
suma2Venn(sigs$summary[, c(1:4)])

# Upset plot of scMaSigPro
plotIntersect(
  scmpObj = scmp.ob,
  min_intersection_size = 0
)
```


### Visualizing Single trend with `maSigPro::PlotGroups()`
The `PlotGroups()` function in `maSigPro` generates a plot of gene expression profiles
with time as the x-axis and gene expression on the y-axis. In these plots, gene
expression from the same experimental group is represented in the same color, and
lines are drawn to join the averages of each time-group to visualize the trend of
each group over time. For `scMaSigPro`, this functionality is achieved using 
`plotTrend()`.
```{r, "STMDE66"}
# Extracting gene "STMDE66" from data
STMDE66 <- data.abiotic[rownames(data.abiotic) == "STMDE66", ]

# Plotting with maSigPro
PlotGroups(STMDE66,
  edesign = edesign.abiotic, show.fit = TRUE,
  dis = design$dis, groups.vector = design$groups.vector
)

# Plotting the same gene with scMaSigPro
plotTrend(scmp.ob, "STMDE66",
  logs = FALSE, pseudoCount = 0,
  smoothness = 1, significant = FALSE
)
```

<!-- ### Visualizing Cluster trend with `maSigPro::PlotProfiles()` -->
<!-- ```{r, "Cluster Plots", fig.width=7} -->
<!-- # Plot clustered Trend -->
<!-- gc <- capture_output( -->
<!--     see.genes(sigs$sig.genes$ColdvsControl, show.fit = T, dis =design$dis, -->
<!-- cluster.method="hclust" , cluster.data = 1, k = 9) -->
<!-- ) -->

<!-- # ScMaSigPro -->
<!-- plotTrendCluster(scmp.ob, -->
<!--                  geneSet = "ColdvsControl", -->
<!--                  logs = F, -->
<!--                  cluster_by = "counts", -->
<!--                  k = 9) -->
<!-- ``` -->

<!-- ```{r, "Cluster Plots2"} -->
<!-- plotTrendCluster(scmp.ob, -->
<!--                  geneSet = "ColdvsControl", -->
<!--                  logs = F, -->
<!--                  cluster_by = "counts", -->
<!--                  k = 9) -->
<!-- ``` -->

---

## Session Info
```{r, "Session Info"}
sessionInfo(package = "scMaSigPro")
sessionInfo(package = "maSigPro")
```
